File: LC/501. Find Mode in Binary Search Tree.py
class Solution:
    def __init__(self):
        self.__cS = 0
        self.__mS = 0
        self.__cN = 0
        self.__ans = []
    def __dfs(self,root):
        if not root:
            return
        self.__dfs(root.left)
        if(root.val == self.__cN):
            self.__cS += 1
        else:
            self.__cS = 1
            self.__cN = root.val
        if self.__cS > self.__mS:
            self.__ans = []
            self.__mS = self.__cS
        if self.__cS == self.__mS:
            self.__ans.append(root.val)
        self.__dfs(root.right)
    def findMode(self, root: Optional[TreeNode]) -> List[int]:
        self.__dfs(root)
        return self.__ans
        

File: LC/501. Find Mode in Binary Search Tree.cpp
class Solution {
private:
    int cS = 0, mS = 0, cN = 0;
    vector<int> ans;
    void dfs(TreeNode* root) {
        if(!root) return;
        dfs(root->left);
        if(cN == root->val) cS++;
        else {
            cS = 1;
            cN = root->val;
        }
        if(cS > mS) {
            ans = {};
            mS = cS;
        }
        if(cS == mS) ans.push_back(root->val);
        dfs(root->right);
    }
public:
    vector<int> findMode(TreeNode* root) {
        dfs(root);
        return ans;
    }
};